docs/C_API_Guidelines.md

```diff
@@ -0,0 +1,83 @@
+# ORT API Guidelines
+
+## Introduction
+
+Our public C API is the main interface of ONNX runtime with our customers. This document endeavors to describe our expectations with respect to the documentation and the quality of the API entry points so we are better equipped to code review new contributions and address existing shortcomings.
+
+The document is entitled guidelines. However, the expectation is that everyone understands it and adheres to it when implementing a new API or while reviewing contributions from others.
+
+## Guidelines
+
+### 1. Public API must be properly documented
+
+All APIs must have a proper documentation header that includes:
+
+* API summary includes any limitations, such as types that it operates on.
+* Description of each of its arguments and whether it is an in, out or in/out argument. Please, document if the user is responsible for memory deallocation or object destruction and how that can be done. Document that strings are UTF-8 encoded.
+* Describe its return value.
+
+Microsoft recommends XML format. Such a format is automatically suggested by Visual Studio when one types 3 consecutive slashes. Both C++ and C# compilers can generate XML documentation with /doc switch as described here. The documentation then can be converted to HTML pages using tools such as Sandcastle.
+
+However, XML format does not seem to be supported in C. Current style looks like Doxygen but the amount of documentation is so small and the amount of effort to document it is big so we might as well change the style of it if we have other suggestions. Otherwise, we need to try and properly document all of it.
```

Comment:
Let's make a decision and document it here. This sentence seems inconclusive.

Reply:
Studio puts XML comments into C# automatically. Let's mandate those there. We can make Doxygen a standard for C/C++ API although internally we can painlessly use XML.

---
In reply to: [517065928](https://github.com/microsoft/onnxruntime/pull/5686#discussion_r517065928) [](ancestors = 517065928)